<a href="https://colab.research.google.com/github/lukmannm/data-science-2026/blob/main/Pertemuan2_Lukman_240401010181.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pertemuan 2 — Eksplorasi Data dengan Pandas & NumPy


- Nama: Lukman Nur Hakim
- NIM: 240401010181
- Kelas: IF401
- Mata Kuliah: Data Science

## Dataset
Titanic


In [ ]:
# Import library Pandas untuk manipulasi dan analisis data
import pandas as pd

# Import library NumPy untuk operasi numerik dan array
import numpy as np

print('Library berhasil diimpor')
print(f'   Pandas  versi: {pd.__version__}')
print(f'   NumPy   versi: {np.__version__}')

Library berhasil diimpor
   Pandas  versi: 2.2.2
   NumPy   versi: 2.0.2


In [ ]:
# Load dataset Titanic
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

df = pd.read_csv(url)

print('Dataset berhasil dimuat')
df.head()

Dataset berhasil dimuat


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Q1. Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?

In [ ]:
# df.shape mengembalikan tuple (baris, kolom)
jumlah_baris, jumlah_kolom = df.shape
print(f'Jumlah Baris : {jumlah_baris}')
print(f'Jumlah Kolom : {jumlah_kolom}')

# Menampilkan nama kolom beserta tipe datanya
print('\nDaftar Kolom yang Tersedia:')
for i, col in enumerate(df.columns, 1):
    # enumerate mulai dari 1 agar nomor urut lebih rapi
    print(f'  {i:2}. {col} ({df[col].dtype})')

Jumlah Baris : 891
Jumlah Kolom : 13

Daftar Kolom yang Tersedia:
   1. PassengerId (int64)
   2. Survived (int64)
   3. Pclass (int64)
   4. Name (object)
   5. Sex (object)
   6. Age (float64)
   7. SibSp (int64)
   8. Parch (int64)
   9. Ticket (object)
  10. Fare (float64)
  11. Cabin (object)
  12. Embarked (object)
  13. AgeGroup (category)


## Q2. Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?

In [ ]:
# isnull() menghasilkan True/False per cell, sum() menjumlahkan True-nya
missing_count = df.isnull().sum()

# Proporsi = jumlah missing dibagi total baris, dikali 100
missing_pct = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({
    'Jumlah Missing': missing_count,
    'Proporsi (%)': missing_pct.round(2)
})

# Filter hanya kolom yang benar-benar punya missing values
missing_df = missing_df[missing_df['Jumlah Missing'] > 0].sort_values('Proporsi (%)', ascending=False)

print('Kolom dengan Missing Values:')
print(missing_df)
print(f'\nTotal kolom dengan missing: {len(missing_df)} dari {df.shape[1]} kolom')

Kolom dengan Missing Values:
          Jumlah Missing  Proporsi (%)
Cabin                687         77.10
Age                  177         19.87
AgeGroup             177         19.87
Embarked               2          0.22

Total kolom dengan missing: 4 dari 13 kolom


## Q3. Berapa persen penumpang yang selamat (Survived=1)? Gunakan operasi NumPy atau Pandas.

In [ ]:
# Konversi kolom Survived ke NumPy array
survived_array = df['Survived'].values

# np.mean() pada array 0/1 langsung menghasilkan proporsi yang selamat
pct_selamat = np.mean(survived_array) * 100
pct_tidak = 100 - pct_selamat

# np.sum() menghitung jumlah penumpang yang selamat (nilai 1)
print(f'Total penumpang      : {len(survived_array)}')
print(f'Selamat (Survived=1) : {np.sum(survived_array == 1)} ({pct_selamat:.2f}%)')
print(f'Tidak selamat (=0)   : {np.sum(survived_array == 0)} ({pct_tidak:.2f}%)')

Total penumpang      : 891
Selamat (Survived=1) : 342 (38.38%)
Tidak selamat (=0)   : 549 (61.62%)


## Q4. Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa ratarata usianya?

In [ ]:
# Kondisi 1: kolom Sex harus 'female'
# Kondisi 2: kolom Pclass harus 1
# Operator & menggabungkan dua kondisi (keduanya harus True)
wanita_kelas1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]

# Menampilkan jumlah hasil filter
print(f'Jumlah wanita kelas 1 : {len(wanita_kelas1)} orang')

# .mean() otomatis mengabaikan NaN saat menghitung rata-rata
print(f'Rata-rata usia        : {wanita_kelas1["Age"].mean():.2f} tahun')

Jumlah wanita kelas 1 : 94 orang
Rata-rata usia        : 34.61 tahun


## Q5. Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?

In [ ]:
# groupby('Pclass') mengelompokkan data berdasarkan kelas tiket (1, 2, 3)
# .mean() pada kolom Survived (0/1) menghasilkan proporsi penumpang yang selamat
print(df.groupby('Pclass')['Survived'].mean())

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


## Q6. Buat analisis tambahan berdasarkan variabel lain yang Anda anggap menarik. Sertakan interpretasi singkat.
Pada analisis ini saya memilih 3 variabel tambahan yang menarik:
1. **Jenis Kelamin (Sex)** — apakah gender mempengaruhi peluang selamat?
2. **Kelas x Gender** — bagaimana kombinasi kelas dan gender mempengaruhi keselamatan?
3. **Kelompok Usia (AgeGroup)** — apakah usia muda lebih diprioritaskan saat evakuasi?

In [ ]:

print('=== Survival Rate by Gender ===')
print(df.groupby('Sex')['Survived'].mean().mul(100).round(2).to_string())


print('\n=== Survival Rate (%) — Kelas x Gender ===')
pivot = df.groupby(['Pclass', 'Sex'])['Survived'].mean().mul(100).round(2).unstack()
pivot.index = ['Kelas 1', 'Kelas 2', 'Kelas 3']
print(pivot.to_string())


df['AgeGroup'] = pd.cut(df['Age'], bins=[0,12,18,35,60,100],
    labels=['Anak', 'Remaja', 'Dewasa Muda', 'Dewasa', 'Lansia'])
print('\n=== Survival Rate (%) — Kelompok Usia ===')
print(df.groupby('AgeGroup', observed=True)['Survived'].mean().mul(100).round(2).to_string())

=== Survival Rate by Gender ===
Sex
female    74.20
male      18.89

=== Survival Rate (%) — Kelas x Gender ===
Sex      female   male
Kelas 1   96.81  36.89
Kelas 2   92.11  15.74
Kelas 3   50.00  13.54

=== Survival Rate (%) — Kelompok Usia ===
AgeGroup
Anak           57.97
Remaja         42.86
Dewasa Muda    38.27
Dewasa         40.00
Lansia         22.73


A. Survival Rate by Gender:
Wanita memiliki tingkat keselamatan jauh lebih tinggi (sekitar 74%) dibanding pria (sekitar 19%).
Hal ini mencerminkan kebijakan evakuasi "women and children first" yang diterapkan awak kapal.

B. Kelas x Gender:
Wanita di semua kelas memiliki survival rate lebih tinggi dari pria di kelas yang sama.
Ini menunjukkan bahwa gender lebih dominan daripada kelas dalam menentukan peluang selamat.

C. Kelompok Usia:
Anak-anak (0–12 tahun) memiliki survival rate tertinggi, mendukung prioritas evakuasi untuk anak-anak.
Kelompok lansia justru memiliki survival rate paling rendah.